### DISCUSSION TITLE
Splitting Next.js API endpoints into separate Vercel functions

### DISCUSSION THREAD

**Discussion by dangreaves** [trpc/trpc#455] [General]
Hey @KATT - love the approach you have taken with this. Proper type sharing in Next.js projects is something I have been trying to achieve for a while and your solution here ticks all the boxes.

One thing which I would like hear your thoughts on is whether or not this will ever be able to split endpoints into their own Next.js API functions, rather than one giant `[trpc]` function. The reason for this is that when deploying to serverless providers like Vercel, they split up each API file into it's own Lambda function.

This has the benefit of keeping the weight of each function low which significantly reduces cold start times, because only the code necessary for that particular endpoint is included in the function.

I'm concerned that as my project grows, the `[trpc]` function will gradually get bigger and bigger, until I end up having some ops issues with the Lambda execution.

Is this something which you have considered? Would love to hear your thoughts.

**Comment by KATT** [comment]
Hey @dangreaves, thanks for opening the discussion!

No, I haven't considered in depth, but my thinking of benefits with one global handler is that:

- No/few cold starts as we always hit the same endpoint
- Fewer open SQL-connections left dangling (known issue with serverless + sql)

I very much doubt that composition of the endpoints will have any significant added cold start time, but haven't tested it thoroughly! 

---

That said, it could be possible to expose your sub-routers as individual functions pretty easily but by jumping through some hoops. Haven't tested, but below should work

- Create sub-router called `posts` with a query called `list`
- Make sure that you have a global `appRouter` where it's all glued together in a global `AppRouter` for inference to work
- Create `/pages/api/trpc/posts/[trpc].tsx`
  - Expose and create a handler for _only_ the router you want to target here _(`trpc.createNextApiHandler({router: postsRouter, /* ... */})`_
  - Try to visit `/api/trpc/posts/list`, see if it works

**Reply by KATT** [reply]
@dangreaves let me know if this works! I haven't consciously designed trpc with this in mind, but it should work, and I'm open to do lil' tweaks to make it possible if not.

**Reply by dangreaves** [reply]
Hey @KATT - thanks so much for breaking that down.

I was considering something along these lines but was concerned that the React hook wouldn't be able to workout the endpoint URL. However, it looks like the way you have structured that means that `posts.list` would be the same as `list` in it's own router, so looks like it will work fine!

Will let you know if I get a chance to test this, but if it works, adding a note like this to the documentation could be cool, to show that it can be done both ways!

Also, agree with your points re. the one big function. It might not actually affect anything at all, but I personally prefer to have them split for logging and monitoring purposes, especially as the API grows bigger and bigger, it's easy to see which endpoints are taking too long, throwing errors etc without digging into one big log group.

**Reply by KATT** [reply]
Great! Let me know how it works, but it feels a bit like premature optimization and there's more than one way of solving the monitoring issue. There are a few future things that won't be possible as well, like query batching (#273)
